src/interactive/repl.ts

```diff
@@ -25,6 +26,18 @@ let g_terminal: vscode.Terminal = null
 
 export let g_connection: rpc.MessageConnection = undefined
 
+export function withConnection(what: string, callback: (connection: rpc.MessageConnection) => void) {
+    if (isConnectionActive()) {
+        callback(g_connection)
+    } else {
+        vscode.window.showInformationMessage(`${what} requires an active REPL.`)
+    }
+}
```

Comment:
Just one random comment here: I think general we probably need to use a different pattern for these kinds of things that takes into account that `g_connection` can easily get closed right after our call to `isConnectionActive` by the other end of the communication channel. As far as I can tell the only way to make these kinds of situations robust is to use `try catch` blocks.

We had a similar situation in the LS: for example we checked whether a file exists on disc, and then opened it. Turns out that with the huge user numbers we had we ran into situations where we checked for the existence of the file, then the file was deleted, and then we tried to open it. So we had to change all of those pieces of code to a pattern where we didn't even bother to check whether the file exists or not, but instead just catch errors if the `open` call fails because that covers both the case that the file didn't exist in the first place _and_ the case that the file gets deleted in the middle of our own code path.

Reply:
maybe addressed by eb5ab9b ?

Reply:
I think that is too broad, as the `catch` clause will now catch any exception in the `callback`, even exceptions that are not due to the connection being closed. I don't think this pattern here with where one passes one fairly large callback will work, I think instead we need a narrow try catch block around every call that sends a request or notification, with narrow error handling for that particular location.